### PR TITLE
parser: improve docs and small implementation changes

### PR DESCRIPTION
**Pull request by lucasem**

### CODE REVIEW

File: src/parser.rs [hunk 1]
```diff
@@ -60,36 +60,60 @@ pub fn parse_term(sig: &mut Signature, input: &str) -> Result<Term, ParseError>
 ///
 /// # TRS syntax
 ///
-/// `input` is parsed as a `<program>`, defined as follows:
+/// The input is parsed as a `program`, defined as follows in [augmented Backus-Naur form]:
```
Comment (joshrule) [inline src/parser.rs]: Why move to augmented BNF?

File: src/parser.rs [hunk 2]
```diff
@@ -60,36 +60,60 @@ pub fn parse_term(sig: &mut Signature, input: &str) -> Result<Term, ParseError>
...
+/// top-level-term /= top-level-term 1*wsp top-level-term
 ///
-/// <top-level-term) ::= ( <term> | ( "(" <top-level-term> ")" ) ) (" "  ( <term> | ( "(" <top-level-term> ")" ) ) )*
+/// term = variable / application
 ///
-/// <term> ::= <variable> | <application>
+/// variable = identifier"_"
 ///
-/// <variable> ::= <identifier>"_"
+/// application = identifier "(" *( *wsp term *wsp ) ")"
```
Comment (joshrule) [inline src/parser.rs]: There *must* be whitespace between subsequent terms in an application, so this definition needs to revised, I think.

File: src/parser.rs (hunk 1 above)
Comment (lucasem) [reply src/parser.rs]: My main motivation was the readability of incremental alternatives (the lines with `left /= alternative`).

File: src/parser.rs (hunk 2 above)
Comment (lucasem) [reply src/parser.rs]: It actually isn't required by the parser. We could though — and probably should. I'll add that.

**Pull request merged**